base_layer/core/src/transaction_protocol_manager.rs

```diff
@@ -0,0 +1,1046 @@
+            _ => Err(TransactionProtocolError::InvalidStateError),
+        }
+    }
+
+    /// The Receiver can receive this public data if they are in the ReceiverOutputSetup state and it will transition
+    /// them to the SenderPartialSignature state
+    pub fn accept_receiver_public_data(
+        &mut self,
+        incoming_data: ReceiverPublicData,
+    ) -> Result<(), TransactionProtocolError>
+    {
+        self.handle_event(TransactionProtocolEvent::AcceptReceiverPublicData(incoming_data))
+    }
+
+    // TransactionProtocolState::SenderPartialSignature state query methods
+    // --------------------------------------------------------------------
+
+    /// Method to determine if we are in the SenderSetup state
+    pub fn is_sender_setup(&self) -> bool {
+        match self.state {
+            TransactionProtocolState::SenderSetup(_) => true,
+            _ => false,
+        }
+    }
+
+    /// Method to determine if we are in the ReceiverOutputSetup state
+    pub fn is_receiver_output_setup(&self) -> bool {
+        match self.state {
+            TransactionProtocolState::ReceiverOutputSetup(_) => true,
+            _ => false,
+        }
+    }
+
+    /// Method to determine if we are in the ReceiverPartialSignature state
+    pub fn is_receiver_partial_signature(&self) -> bool {
+        match self.state {
+            TransactionProtocolState::ReceiverPartialSignature(_) => true,
+            _ => false,
+        }
+    }
+
+    /// Method to determine if we are in the SenderPartialSignature state
+    pub fn is_sender_partial_signature(&self) -> bool {
+        match self.state {
+            TransactionProtocolState::SenderPartialSignature(_) => true,
+            _ => false,
+        }
+    }
+
+    /// Method to determine if we are in the FinalizedTransaction state
+    pub fn is_finalized(&self) -> bool {
+        match self.state {
+            TransactionProtocolState::FinalizedTransaction(_) => true,
+            _ => false,
+        }
+    }
+
+    /// Method to determine if we are in the Failed state
+    pub fn is_failed(&self) -> bool {
+        match self.state {
+            TransactionProtocolState::Failed(_) => true,
+            _ => false,
+        }
+    }
+
+    /// Method to return the error behind a failure, if one has occured
+    pub fn failure_reason(&self) -> Option<TransactionProtocolError> {
+        match &self.state {
+            TransactionProtocolState::Failed(e) => Some(e.clone()),
+            _ => None,
+        }
+    }
+
+    /// This function implements the state machine for the Transaction Protocol. Every combination of State and Event
+    /// are handled here. The previous state is consumed and dependant on the outcome of processing the event a new
+    /// state is returned.
+    fn handle_event(&mut self, event: TransactionProtocolEvent) -> Result<(), TransactionProtocolError> {
+        self.state = match self.state.clone() {
```

Comment:
Seems strange to clone it here and then overwrite it. Could we perhaps use a `ref` in the match arms?

Reply:
So I tried many combo's but the issue is that the state changes consume the old state and give you a new one so I couldn't figure out how to allow that but still assign the new state to main struct without having move errors. It does feel a little redundant to make the state changes consume the old state if I am cloning here...

Reply:
Agree on the code smell here. you can have `&mut self` and then try burn the state in the arms.